# Importing packages

In [1]:
import numpy as np
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Setting up the data

Setting up a .csv file in a pandas' dataframe and creating a connection with PostgreSQL. Then I'll convert it to SQL table, so I can query the database through Python & Jupyter Notebooks:

In [3]:
df = pd.read_csv('valid_covid_19_data.csv', dtype={'intubed':'string', 'pregnant':'string', 'icu':'string'}, parse_dates=[4])
engine = create_engine(f'postgresql+psycopg2://postgres:{password}@localhost:5432/case_study_2')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 21 columns):
 #   Column                Non-Null Count    Dtype         
---  ------                --------------    -----         
 0   treatment_type        1048575 non-null  int64         
 1   medical_unit          1048575 non-null  int64         
 2   sex                   1048575 non-null  object        
 3   care_received         1048575 non-null  object        
 4   date_died             76942 non-null    datetime64[ns]
 5   intubed               192706 non-null   string        
 6   pneumonia             1032572 non-null  object        
 7   age                   1048575 non-null  int64         
 8   pregnant              521310 non-null   string        
 9   diabetes              1045237 non-null  object        
 10  copd                  1045572 non-null  object        
 11  asthma                1045596 non-null  object        
 12  immunosuppressed      1045171 non-null  ob

In [ ]:
sql_df = df.to_sql('valid_covid_19_data', engine) 

Table is set up correctly now. To begin, let's see if the dataset values weren't somehow mixed/altered in the process.
Here's the preview of copied table and it's columns:

# Analysis

This will be more exploratory than in-depth analysis, so the queries will guide you through some of the most important insights the dataset provides.

First, let's see the small sample of the data:

In [6]:
pd.read_sql_query("""
    SELECT * FROM valid_covid_19_data
    LIMIT 50
""", engine)

treatment_type  medical_unit sex  care_received   date_died intubed  \
0                2             1   F  Returned home  2020-05-03    None   
1                2             3   F  Returned home  2020-04-20    None   
2                2             3   F  Returned home  2020-05-13    None   
3                2             3   F  Returned home  2020-05-20    None   
4                2             3   F  Returned home  2020-05-24    None   
5                2             3   F  Returned home  2020-05-25    None   
6                1             3   F  Returned home  2020-05-27    None   
7                2             3   F  Returned home  2020-06-03    None   
8                1             3   F  Returned home  2020-06-06    None   
9                2             3   F  Returned home  2020-06-09    None   
10               2             3   F  Returned home  2020-06-09    None   
11               2             3   F  Returned home  2020-06-15    None   
12               2             3   F  Returned home  2020-06-16    None   
13               2             3   F  Returned home  2020-06-16    None   
14               2             3   F  Returned home  2020-06-17    None   
15               2             3   F  Returned home  2020-06-17    None   
16               2             3   F  Returned home  2020-06-19    None   
17               2             3   F  Returned home  2020-06-23    None   
18               1             3   F  Returned home  2020-06-25    None   
19               2             3   F  Returned home  2020-06-28    None   
20               2             3   F  Returned home  2020-06-30    None   
21               1             3   F  Returned home  2020-07-03    None   
22               2             3   F  Returned home  2020-07-05    None   
23               1             3   F  Returned home  2020-07-06    None   
24               1             3   F  Returned home  2020-07-12    None   
25               2             3   F  Returned home  2020-07-13    None   
26               1             3   F  Returned home  2020-07-16    None   
27               2             3   F  Returned home  2020-07-19    None   
28               2             3   F  Returned home  2020-07-30    None   
29               1             3   F  Returned home  2020-08-03    None   
30               2             3   F  Returned home  2020-08-06    None   
31               2             3   F  Returned home  2020-08-10    None   
32               1             4   F  Returned home  2020-04-08    None   
33               2             4   F  Returned home  2020-04-08    None   
34               1             4   F  Returned home  2020-04-08    None   
35               2             4   F  Returned home  2020-04-13    None   
36               2             4   F  Returned home  2020-04-13    None   
37               2             4   F  Returned home  2020-04-14    None   
38               2             4   F  Returned home  2020-04-14    None   
39               2             4   F  Returned home  2020-04-15    None   
40               1             4   F  Returned home  2020-04-15    None   
41               2             4   F  Returned home  2020-04-16    None   
42               1             4   F  Returned home  2020-04-17    None   
43               1             4   F  Returned home  2020-04-17    None   
44               2             4   F  Returned home  2020-04-18    None   
45               1             4   F  Returned home  2020-04-18    None   
46               1             4   F  Returned home  2020-04-20    None   
47               2             4   F  Returned home  2020-04-21    None   
48               2             4   F  Returned home  2020-04-21    None   
49               1             4   F  Returned home  2020-04-21    None   

    pneumonia  age  pregnant  diabetes   copd  asthma  immunosuppressed  \
0        True   65     False     False  False   False             False   
1        True   71     False

**Note:** It appears that null values were replaced with 'None' values. It will not affect the output of sql queries, so I'm not going to change these values.

Let's see how many unique patients are registered in the dataset:

In [7]:
pd.read_sql_query("""
    SELECT COUNT(*) all_patients
    FROM valid_covid_19_data
""", engine)

all_patients
0       1048575

Time period covered in the dataset (starting date, ending date, and amount of days between them):

In [8]:
pd.read_sql_query("""
    SELECT 
        MIN(date_died) starting_date, 
        MAX(date_died) ending_date, 
        MAX(date_died) - MIN(date_died) days_span
    FROM valid_covid_19_data
""", engine)

starting_date ending_date  days_span
0    2020-01-02  2021-05-02        486

Patients distribution by sex:

In [9]:
pd.read_sql_query("""
    SELECT 
        sex, 
        COUNT(sex)
    FROM covid_19_table
    GROUP BY sex
""", engine)

sex   count
0   F  525064
1   M  523511

Difference is almost unnoticeable, slightly over 1500 more female patients than male ones. Equal samples from both groups are always good in terms of any research or analysis (especially when they're compared against each other).

Count of all patients, deaths, and patients death rate:                           

In [36]:
pd.read_sql_query("""
    SELECT
        COUNT(*) patients, 
        COUNT(date_died) deaths, 
        CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(*)::numeric, 2) || '%%') death_rate
    FROM valid_covid_19_data 
""", engine)

patients  deaths death_rate
0   1048575   76942      7.34%

Let's divide it by gender:

In [33]:
pd.read_sql_query("""
    SELECT
        sex,
        COUNT(*) all_patients,
        COUNT(date_died) total_deaths,  
        CONCAT(ROUND(100 * COUNT(date_died) / COUNT(*)::numeric,2) || '%%') as death_rate
    FROM valid_covid_19_data
    GROUP BY sex
""", engine)

sex  all_patients  total_deaths death_rate
0   M        523511         49540      9.46%
1   F        525064         27402      5.22%

As we can see, male mortality was much higher, that's close to a twice as much deaths among men than women.

Intubation rate by covid test result:

In [12]:
pd.read_sql_query("""
    SELECT 
        COUNT(*) all_patients, 
        classification_final as test_result,
        COUNT(intubed) intubed_patients, 
        CONCAT(ROUND(COUNT(intubed) * 100 / COUNT(*)::numeric, 2) || '%%') intubation_rate
    FROM valid_covid_19_data
    GROUP BY classification_final
""", engine)

all_patients                      test_result  intubed_patients  \
0        391979                            COVID            109779   
1        656596  Test inconclusive/not a carrier             82927   

  intubation_rate  
0          28.01%  
1          12.63%

Data shows that almost 30% of patients had respiratory problems when dealing with confirmed COVID. That number is less than half of it, when COVID was not confirmed. Considering that majority of patients haven't had COVID confirmed, we might say that COVID can affect respiration.

Deaths per month ordered descending by deaths:

In [14]:
pd.read_sql_query("""
    SELECT 
        TO_CHAR(date_died, 'MM-YYYY') months, 
        COUNT(date_died) deaths_per_month
    FROM valid_covid_19_data
    WHERE date_died IS NOT NULL
    GROUP BY months
    ORDER BY deaths_per_month DESC
""", engine)

months  deaths_per_month
0   06-2020             25278
1   05-2020             21314
2   07-2020             20457
3   04-2020              6926
4   08-2020              1134
5   03-2020               552
6   02-2020               437
7   01-2020               368
8   04-2021               170
9   09-2020                91
10  05-2021                49
11  02-2021                35
12  03-2021                32
13  01-2021                31
14  10-2020                27
15  12-2020                23
16  11-2020                18

We can see that during the 05-2020 to 07-2020 period vast majority of deaths happened. It was in the middle of first wave of COVID around the world, the wave in which people were most vulnerable to this disease. Vaccines were in use in the end and after that phase, so there's no telling how much of these deaths could've been avoided. 

Only in Mexico there were over 70000 deaths in the span of 3 months. Just wondering about the number of victims in this period in the whole world is truly terrifying.

Let's check it even further. 

Deaths per day:

In [15]:
pd.read_sql_query("""
    SELECT 
        date_died, 
        COUNT(date_died) as deaths_per_day
    FROM valid_covid_19_data
    WHERE date_died IS NOT NULL
    GROUP BY date_died
""", engine)

date_died  deaths_per_day
0    2020-01-02               1
1    2020-01-03               5
2    2020-01-04               1
3    2020-01-05               2
4    2020-01-06               3
5    2020-01-07               5
6    2020-01-08               7
7    2020-01-09               7
8    2020-01-10              16
9    2020-01-11               6
10   2020-01-12              10
11   2020-01-13              12
12   2020-01-14              11
13   2020-01-15              22
14   2020-01-16              14
15   2020-01-17              19
16   2020-01-18              23
17   2020-01-19              12
18   2020-01-20              22
19   2020-01-21              13
20   2020-01-22              18
21   2020-01-23              16
22   2020-01-24              17
23   2020-01-25               8
24   2020-01-26              10
25   2020-01-27              16
26   2020-01-28              18
27   2020-01-29              23
28   2020-01-30              15
29   2020-01-31              16
30   2020-02-01              13
31   2020-02-02              15
32   2020-02-03               8
33   2020-02-04              15
34   2020-02-05              14
35   2020-02-06              15
36   2020-02-07              17
37   2020-02-08              14
38   2020-02-09              11
39   2020-02-10              19
40   2020-02-11              22
41   2020-02-12              18
42   2020-02-13              14
43   2020-02-14              11
44   2020-02-15               9
45   2020-02-16              18
46   2020-02-17              18
47   2020-02-18              19
48   2020-02-19              18
49   2020-02-20              21
50   2020-02-21              17
51   2020-02-22              10
52   2020-02-23               8
53   2020-02-24              20
54   2020-02-25              16
55   2020-02-26              20
56   2020-02-27              12
57   2020-02-28              14
58   2020-02-29              11
59   2020-03-01              11
60   2020-03-02              11
61   2020-03-03              14
62   2020-03-04              13
63   2020-03-05               9
64   2020-03-06              17
65   2020-03-07              10
66   2020-03-08               5
67   2020-03-09              13
68   2020-03-10              12
69   2020-03-11              11
70   2020-03-12              12
71   2020-03-13              15
72   2020-03-14               6
73   2020-03-15              10
74   2020-03-16               5
75   2020-03-17               5
76   2020-03-18              15
77   2020-03-19               7
78   2020-03-20              16
79   2020-03-21               8
80   2020-03-22              10
81   2020-03-23              13
82   2020-03-24              22
83   2020-03-25              18
84   2020-03-26              31
85   2020-03-27              36
86   2020-03-28              39
87   2020-03-29              33
88   2020-03-30              50
89   2020-03-31              75
90   2020-04-01              56
91   2020-04-02              67
92   2020-04-03              69
93   2020-04-04              77
94   2020-04-05              68
95   2020-04-06              95
96   2020-04-07             125
97   2020-04-08             104
98   2020-04-09             126
99   2020-04-10             156
100  2020-04-11             150
101  2020-04-12             150
102  2020-04-13             182
103  2020-04-14             190
104  2020-04-15             175
105  2020-04-16             185
106  2020-04-17             209
107  2020-04-18             211
108  2020-04-19             257
109  2020-04-20             292
110  2020-04-21             332
111  2020-04-22             302
112  2020-04-23             335
113  2020-04-24             346
114  2020-04-25             399
115  2020-04-26             418
116  2020-04-27             400
117  2020-04-28             469
118  2020-04-29             509
119  2020-04-30             472
120  2020-05-01             555
121  2020-05-02             523
122  2020-05-03             517
123  2020-05-04             641
124  2

Above query result shows how COVID got to it's peak in previously mentioned period.

Based on the query where I calculated the period covered in the dataset, let's see how many days have gone by without a single death:

In [39]:
pd.read_sql_query("""
    WITH cte AS (
        SELECT 
            date_died, 
            COUNT(date_died) as deaths_per_day
        FROM valid_covid_19_data
        WHERE date_died IS NOT NULL
        GROUP BY date_died)
    SELECT 486 - COUNT(*) days_without_deaths
    FROM cte
""", engine)

days_without_deaths
0                   86

That's actually more than I expected, the ending of 2020 and beginning of 2021 were quite calm in that regard, situation with COVID was slowing down (until another wave came in, sadly). Missing dates in previous query tell us, when these days were.

Count of all patients carrying a certain disease with count of how many of these carriers died (2nd count automatically throws out null values, giving the accurate deaths number) matched with average age of the deceased patients. 
Other_diseases column gives basically random information, so as to not taking guesses what those might be, I've decided to exclude them from the query.

In [16]:
pd.read_sql_query("""
    SELECT 
        'Pneumonia' disease, 
        COUNT(pneumonia) patients, 
        COUNT(pneumonia) - COUNT(date_died) alive, 
        COUNT(date_died) deceased, CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(pneumonia)::numeric, 2) || '%%') disease_death_rate, 
        ROUND(AVG(age),2) average_suffering_patient_age
    FROM valid_covid_19_data
    WHERE pneumonia = True
    GROUP BY pneumonia
    UNION 
    SELECT 
        'Obesity', 
        COUNT(obesity), COUNT(obesity) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(obesity)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE obesity = True
    GROUP BY obesity
    UNION
    SELECT 
        'Diabetes', 
        COUNT(diabetes), COUNT(diabetes) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(diabetes)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE diabetes = True
    GROUP BY diabetes
    UNION
    SELECT 
        'COPD', 
        COUNT(copd), COUNT(copd) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(copd)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE copd = True
    GROUP BY copd
    UNION
    SELECT 
        'Asthma', 
        COUNT(asthma), 
        COUNT(asthma) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(asthma)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE asthma = True
    GROUP BY asthma
    UNION
    SELECT 
        'Immunosuppression', 
        COUNT(immunosuppressed), 
        COUNT(immunosuppressed) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(immunosuppressed)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE immunosuppressed = True
    GROUP BY immunosuppressed
    UNION
    SELECT 
        'Hypertension', 
        COUNT(hypertension), 
        COUNT(hypertension) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(hypertension)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE hypertension = True
    GROUP BY hypertension
    UNION
    SELECT 
        'Cardiovascular', 
        COUNT(cardiovascular), 
        COUNT(cardiovascular) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(cardiovascular)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE cardiovascular = True
    GROUP BY cardiovascular
    UNION
    SELECT 
        'Renal Chronic', 
        COUNT(renal_chronic), 
        COUNT(renal_chronic) - COUNT(date_died), 
        COUNT(date_died), CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(renal_chronic)::numeric, 2) || '%%'), 
        ROUND(AVG(age),2)
    FROM valid_covid_19_data
    WHERE renal_chronic = True
    GROUP BY renal_chronic
""", engine)

disease  patients   alive  deceased disease_death_rate  \
0     Cardiovascular     20769   16334      4435             21.35%   
1      Renal Chronic     18904   13197      5707             30.19%   
2          Pneumonia    140038   86115     53923             38.51%   
3           Diabetes    124989   96724     28265             22.61%   
4  Immunosuppression     14170   11552      2618             18.48%   
5            Obesity    159816  142522     17294             10.82%   
6       Hypertension    162729  130668     32061             19.70%   
7             Asthma     31572   30092      1480              4.69%   
8               COPD     15062   11041      4021             26.70%   

   average_suffering_patient_age  
0                          57.39  
1                          54.41  
2                          53.81  
3                          56.96  
4                          46.50  
5                          45.35  
6                          57.19  
7                          39.27  
8                          65.16

Data tells us, that almost 40% of patients who had pneumonia eventually had passed, the number is extremely high. Pneumonia and renal chronic disease appear to be significantly more dangerous combined with COVID-19 than other diseases. 

The average age for deceased patients who had pneumonia is 53.81 yrs old. I've checked in case it's tied to old age of these patients, but slightly less than 54 years.


Tobacco users by gender:

In [17]:
pd.read_sql_query("""
    SELECT 
        sex, 
        COUNT(*) tobacco_users, 
        (SELECT COUNT(*) FROM valid_covid_19_data) all_patients_count
    FROM valid_covid_19_data
    WHERE tobacco = TRUE 
    GROUP BY sex
""", engine)

sex  tobacco_users  all_patients_count
0   F          27718             1048575
1   M          56658             1048575

Obese patients by gender:

In [18]:
pd.read_sql_query("""
    SELECT 
        sex,
        COUNT(*) as obese_patients
    FROM valid_covid_19_data
    WHERE obesity = TRUE
    GROUP BY 1    
""", engine)

sex  obese_patients
0   F           83424
1   M           76392

Count of positive/negative test results by age.

In [19]:
pd.read_sql_query("""
    SELECT 
        age, 
        classification_final, 
        COUNT(classification_final) results
    FROM valid_covid_19_data
    GROUP BY age, classification_final
    ORDER BY age
""", engine)

age             classification_final  results
0      0                            COVID      744
1      0  Test inconclusive/not a carrier     3118
2      1                            COVID      578
3      1  Test inconclusive/not a carrier     4224
4      2                            COVID      360
5      2  Test inconclusive/not a carrier     2818
6      3                            COVID      318
7      3  Test inconclusive/not a carrier     2241
8      4                            COVID      329
9      4  Test inconclusive/not a carrier     2156
10     5                            COVID      302
11     5  Test inconclusive/not a carrier     2083
12     6                            COVID      366
13     6  Test inconclusive/not a carrier     2044
14     7                            COVID      351
15     7  Test inconclusive/not a carrier     2054
16     8                            COVID      406
17     8  Test inconclusive/not a carrier     2038
18     9                            COVID      393
19     9  Test inconclusive/not a carrier     1956
20    10                            COVID      462
21    10  Test inconclusive/not a carrier     2028
22    11                            COVID      522
23    11  Test inconclusive/not a carrier     2013
24    12                            COVID      562
25    12  Test inconclusive/not a carrier     2007
26    13                            COVID      675
27    13  Test inconclusive/not a carrier     2248
28    14                            COVID      734
29    14  Test inconclusive/not a carrier     2342
30    15                            COVID      872
31    15  Test inconclusive/not a carrier     2645
32    16                            COVID     1002
33    16  Test inconclusive/not a carrier     2921
34    17                            COVID     1190
35    17  Test inconclusive/not a carrier     3302
36    18                            COVID     1354
37    18  Test inconclusive/not a carrier     3956
38    19                            COVID     2021
39    19  Test inconclusive/not a carrier     5941
40    20                            COVID     2549
41    20  Test inconclusive/not a carrier     7383
42    21                            COVID     3106
43    21  Test inconclusive/not a carrier     8181
44    22                            COVID     3423
45    22  Test inconclusive/not a carrier     9266
46    23                            COVID     4358
47    23  Test inconclusive/not a carrier    10957
48    24                            COVID     5311
49    24  Test inconclusive/not a carrier    12935
50    25                            COVID     6248
51    25  Test inconclusive/not a carrier    14792
52    26                            COVID     7169
53    26  Test inconclusive/not a carrier    16042
54    27                            COVID     7620
55    27  Test inconclusive/not a carrier    16714
56    28                            COVID     7979
57    28  Test inconclusive/not a carrier    17334
58    29                            COVID     8240
59    29  Test inconclusive/not a carrier    16894
60    30                            COVID     8899
61    30  Test inconclusive/not a carrier    18111
62    31                            COVID     8600
63    31  Test inconclusive/not a carrier    17327
64    32                            COVID     8391
65    32  Test inconclusive/not a carrier    16470
66    33                            COVID     8567
67    33  Test inconclusive/not a carrier    16062
68    34                            COVID     8688
69    34  Test inconclusive/not a carrier    16184
70    35                            COVID     8733
71    35  Test inconclusive/not a carrier    15902
72    36                            COVID     8711
73    36  Test inconclusive/not a carrier    15887
74    37                            COVID     9019
75    37  Test inconclusive/not a carrier    15638
76    38                            COVID     8811
77    38  Test inconclusive

As such queries go, people with age around mean age of all the patients are the most frequent ones, so it had to be them who had the highest number in positive COVID test results. They've the highest number, but the highest percentage belongs to slightly older to old people (50-80 yrs old). Human's immune system starts to fail more often with age, so virus presence makes them even more vulnerable or susceptible to succumbing to COVID or other serious diseases. As we can see above people with the age of 65-70 had almost the same numbers when divided among negative and positive results, and when we'll take a look at the groups of 30-35, the ratio is much lower. 

Minimum, maximum and average age for people with confirmed COVID and those without divided by sex:

In [20]:
pd.read_sql_query("""
    SELECT 
        'COVID' as test_result,
        sex, 
        MIN(age) minimum_age, 
        MAX(age) maximum_age, 
        ROUND(AVG(age),2) average_age
    FROM valid_covid_19_data
    WHERE classification_final = 'COVID'
    GROUP BY sex
    UNION
    SELECT 
        'Inconclusive/not a carrier', 
        sex, 
        MIN(age) minimum_age, 
        MAX(age) maximum_age, 
        ROUND(AVG(age),2) average_age
    FROM valid_covid_19_data
    WHERE classification_final = 'Inconclusive/not a carrier'
    GROUP BY sex
    ORDER BY sex
""", engine)

test_result sex  minimum_age  maximum_age  average_age
0       COVID   F            0          119        44.42
1       COVID   M            0          120        45.86

Data tells us that oldest tested patients were ~120 yrs old. That number is especially (and suspiciously) high knowing that only one person in the history was confirmed to have lived over 120 years (source:(https://grg.org/WSRL/TableE.aspx))

Having that in mind, that data is most likely falsified, but since I can't be 100% certain about it, I'm not gonna leave it out of the dataset.

Stationed & deceased patients per medical unit ordered by total patients descending (rounded up percentages):

In [21]:
pd.read_sql_query("""
    SELECT 
        medical_unit, 
        COUNT(date_died) deceased, 
        COUNT(*) patients_per_med_unit, 
        CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(*)::numeric, 2) || '%%') deceased_percentage
    FROM valid_covid_19_data 
    GROUP BY medical_unit
    ORDER BY patients_per_med_unit DESC
""", engine)

medical_unit  deceased  patients_per_med_unit deceased_percentage
0             12     24620                 602995               4.08%
1              4     39905                 314405              12.69%
2              6      5790                  40584              14.27%
3              9      1369                  38116               3.59%
4              3      1492                  19175               7.78%
5              8      1171                  10399              11.26%
6             10      1468                   7873              18.65%
7              5       607                   7244               8.38%
8             11       409                   5577               7.33%
9             13        61                    996               6.12%
10             7        40                    891               4.49%
11             2         5                    169               2.96%
12             1         5                    151               3.31%

There are big disproportions between units 4 and 12 and the rest of them. That's probably connected to the localization of these units. Data provided can't answer that question, though.

How were people taken care of and percentage of mortality (rounded up percentages):

In [22]:
pd.read_sql_query("""
    SELECT 
        care_received, 
        COUNT(care_received) people, 
        COUNT(date_died) deaths, 
        CONCAT(ROUND(COUNT(date_died) * 100 / COUNT(care_received)::numeric, 2) || '%%') death_rate
    FROM valid_covid_19_data
    GROUP BY care_received
""", engine)

care_received  people  deaths death_rate
0   Hospitalized  200031   70066     35.03%
1  Returned home  848544    6876      0.81%

There's a very big difference there. It shows how serious the complications after (and through) COVID are. If one of the patients from the dataset were to be hospitalized, there were over 1/3 chance they'd eventually die, which is staggering. We might say, that if the immune system couldn't defend the organism, there was a big chance doctors & medicine couldn't help too.

Deaths by age groups, ordered by number of deaths (I'll consider the age issue here and not include people who were more than 100 years old at the moment:

In [23]:
pd.read_sql_query("""
    SELECT 
        sex,
        ROUND(AVG(age), 0)::int age, 
        COUNT(date_died) deaths
    FROM valid_covid_19_data
    WHERE date_died IS NOT NULL
    GROUP BY age, sex
    HAVING age < 100
    ORDER BY deaths DESC
""", engine)

sex  age  deaths
0     M   60    1380
1     M   65    1370
2     M   56    1365
3     M   61    1363
4     M   59    1333
5     M   63    1311
6     M   66    1301
7     M   64    1278
8     M   68    1264
9     M   67    1262
10    M   57    1254
11    M   62    1251
12    M   58    1239
13    M   55    1222
14    M   70    1199
15    M   69    1190
16    M   53    1150
17    M   52    1118
18    M   71    1107
19    M   72    1098
20    M   54    1066
21    M   50    1028
22    M   73     974
23    M   51     964
24    M   74     956
25    M   48     886
26    M   75     877
27    M   49     847
28    M   76     839
29    M   47     839
30    F   66     812
31    M   77     803
32    M   46     784
33    M   78     783
34    F   61     772
35    F   65     770
36    F   68     763
37    F   63     755
38    F   67     749
39    M   45     744
40    F   62     725
41    F   60     723
42    M   80     701
43    F   70     697
44    F   64     696
45    F   59     682
46    F   57     673
47    F   56     664
48    F   58     664
49    F   71     663
50    F   69     661
51    F   72     657
52    M   44     627
53    M   79     625
54    F   73     606
55    F   75     585
56    F   74     581
57    M   81     577
58    F   55     566
59    M   43     565
60    M   42     562
61    F   78     551
62    F   54     548
63    M   82     542
64    F   77     529
65    F   76     528
66    F   52     514
67    F   53     510
68    M   83     478
69    M   41     473
70    F   51     473
71    F   80     464
72    M   40     449
73    M   84     447
74    F   49     434
75    F   50     414
76    F   79     411
77    F   48     391
78    M   38     391
79    F   81     372
80    M   39     371
81    F   83     361
82    M   85     360
83    F   82     360
84    F   47     356
85    F   46     341
86    F   84     337
87    F   45     318
88    M   87     314
89    M   86     312
90    M   36     309
91    M   37     306
92    F   44     290
93    M   35     269
94    F   85     264
95    M   34     259
96    F   43     234
97    M   33     225
98    M   31     221
99    F   42     220
100   F   86     219
101   F   87     216
102   M   88     215
103   M   32     199
104   F   41     195
105   M   89     193
106   F   88     188
107   F   40     178
108   M   30     177
109   F   38     174
110   F   39     171
111   M   90     164
112   F   90     159
113   M   29     136
114   F   37     135
115   M    0     132
116   M   27     126
117   F   89     123
118   M   28     119
119   F   36     113
120   M   26     112
121   F   34     107
122   F   35     103
123   M   91     103
124   F   33     102
125   M    1     101
126   F   30      96
127   F   91      92
128   F   32      90
129   F   31      88
130   M   25      85
131   F    1      84
132   F   29      79
133   F    0      77
134   M   93      76
135   M   92      74
136   F   28      69
137   F   92      67
138   M   24      65
139   F   26      64
140   F   93      59
141   F   27      53
142   M   21      53
143   M   22      50
144   M   94      48
145   F   25      47
146   M   95      47
147   F   94      46
148   M   23      43
149   F   24      43
150   F   22      39
151   F   23      35
152   F   95      34
153   M   20      32
154   F   96      31
155   M    2      31
156   F   21      30
157   M   96      28
158   M   97      27
159   F   20      26
160   F    2      24
161   F   97      22
162   M   19      20
163   F   19      20
164   M   14      19
165   F   17      19
166   F   98      18
167   M    7      18
168   F   18      18
169   M   98      18
170   M   15      18
171   M    6      17
172   M    3      16
173   F   11      15
174   M   17      15
175   M   16      14
176   F    3      14
177   M   18      14
178   F   15      13
179   M   10      13
180   F   10      13
181   F    5      12
182   F    9      11
183   M   99      11
184   M   12      11
185   F    8      10
186   F   14      10
187   M   13      10
188   M   11      10
189   M    8  

Overall, the 30th age-group is the first one consisting of females. Considering the majority (small one though) of the examined patients were females, it's extremely alarming for males. That might be due to their lifestyle, working conditions and many other aspects but above all - their age (which as I previously mentioned affects the immune system) and we can see that these groups don't consist of the youngest people. Not necessarily only the gender itself plays it's part here.

Count of deaths and survivors of those who had an ICU intervention:

In [24]:
pd.read_sql_query("""
    SELECT 
        'Dead after ICU intervention' as outcome, 
        COUNT(icu)
    FROM valid_covid_19_data
    WHERE icu = TRUE and date_died IS NOT NULL
    UNION
    SELECT 
        'Alive after ICU intervention', 
        COUNT(icu)
    FROM valid_covid_19_data
    WHERE icu = TRUE and date_died IS NULL
""", engine)

outcome  count
0  Alive after ICU intervention   8663
1   Dead after ICU intervention   8195

ICU intervention means that situation is extremely dangerous - based on hospitalised patients death rate I expected the ratio here to be higher on the side of deaths number. Luckily, many of those people made it out alive.

Let's see which day was the peak one in terms of deaths:

In [25]:
pd.read_sql_query("""
    SELECT 
        date_died, 
        COUNT(date_died)
    FROM valid_covid_19_data
    GROUP BY date_died
    ORDER BY COUNT(date_died) DESC
    LIMIT 1
""", engine)

date_died  count
0  2020-07-06   1000

1000 deaths in one day on 6th of July 2020. In the middle of the first international COVID wave, very likely the most deadly period since COVID took over the globe.

On the other side, let's see which days had the least deceased patients, sorted chronologically and by a number of deceased patients:

In [26]:
pd.read_sql_query("""
    SELECT 
        date_died, 
        COUNT(date_died)
    FROM valid_covid_19_data
    WHERE date_died IS NOT NULL 
    GROUP BY date_died
    HAVING COUNT(date_died) <= 10
    ORDER BY COUNT(date_died), date_died ASC
""", engine)

date_died  count
0    2020-01-02      1
1    2020-01-04      1
2    2020-09-06      1
3    2020-09-10      1
4    2020-09-23      1
5    2020-09-25      1
6    2020-10-07      1
7    2020-10-10      1
8    2020-10-12      1
9    2020-10-13      1
10   2020-10-16      1
11   2020-10-17      1
12   2020-10-18      1
13   2020-10-20      1
14   2020-10-22      1
15   2020-10-23      1
16   2020-10-27      1
17   2020-10-30      1
18   2020-10-31      1
19   2020-11-06      1
20   2020-11-15      1
21   2020-11-22      1
22   2020-11-24      1
23   2020-11-25      1
24   2020-11-27      1
25   2020-12-03      1
26   2020-12-05      1
27   2020-12-08      1
28   2020-12-16      1
29   2020-12-17      1
30   2020-12-19      1
31   2020-12-20      1
32   2020-12-22      1
33   2020-12-24      1
34   2020-12-25      1
35   2020-12-30      1
36   2021-01-01      1
37   2021-01-02      1
38   2021-01-06      1
39   2021-01-07      1
40   2021-01-10      1
41   2021-01-11      1
42   2021-01-14      1
43   2021-01-19      1
44   2021-01-25      1
45   2021-02-01      1
46   2021-02-07      1
47   2021-02-10      1
48   2021-02-11      1
49   2021-02-14      1
50   2021-02-19      1
51   2021-02-20      1
52   2021-02-22      1
53   2021-02-23      1
54   2021-02-26      1
55   2021-02-27      1
56   2021-03-07      1
57   2021-03-10      1
58   2021-03-11      1
59   2021-03-16      1
60   2021-03-19      1
61   2021-03-24      1
62   2021-03-26      1
63   2021-03-27      1
64   2021-03-31      1
65   2021-04-02      1
66   2021-04-13      1
67   2021-04-19      1
68   2021-04-20      1
69   2021-04-22      1
70   2020-01-05      2
71   2020-09-12      2
72   2020-09-15      2
73   2020-09-18      2
74   2020-09-24      2
75   2020-09-27      2
76   2020-09-29      2
77   2020-10-04      2
78   2020-10-06      2
79   2020-10-14      2
80   2020-10-26      2
81   2020-11-12      2
82   2020-11-14      2
83   2020-11-23      2
84   2020-12-07      2
85   2020-12-15      2
86   2020-12-18      2
87   2020-12-28      2
88   2021-01-15      2
89   2021-01-22      2
90   2021-01-24      2
91   2021-01-26      2
92   2021-01-28      2
93   2021-01-31      2
94   2021-02-04      2
95   2021-02-05      2
96   2021-02-06      2
97   2021-02-08      2
98   2021-02-12      2
99   2021-02-24      2
100  2021-02-25      2
101  2021-03-08      2
102  2021-03-14      2
103  2021-03-20      2
104  2021-03-22      2
105  2021-03-28      2
106  2021-03-29      2
107  2021-03-30      2
108  2021-04-01      2
109  2021-04-06      2
110  2021-04-09      2
111  2021-04-10      2
112  2021-04-15      2
113  2021-04-16      2
114  2021-04-17      2
115  2021-04-18      2
116  2020-01-06      3
117  2020-08-30      3
118  2020-09-01      3
119  2020-09-04      3
120  2020-09-07      3
121  2020-09-08      3
122  2020-09-14      3
123  2020-09-16      3
124  2020-09-17      3
125  2020-09-20      3
126  2020-09-22      3
127  2020-10-05      3
128  2020-10-25      3
129  2020-11-04      3
130  2020-11-08      3
131  2021-01-20      3
132  2021-01-27      3
133  2021-02-03      3
134  2021-02-18      3
135  2021-03-17      3
136  2021-03-18      3
137  2021-03-23      3
138  2021-04-12      3
139  2021-04-14      3
140  2020-09-11      4
141  2020-09-13      4
142  2020-09-19      4
143  2020-09-21      4
144  2020-09-28      4
145  2020-12-26      4
146  2021-01-21      4
147  2021-02-17      4
148  2021-04-03      4
149  2021-04-24      4
150  2020-01-03      5
151  2020-01-07      5
152  2020-03-08      5
153  2020-03-16      5
154  2020-03-17      5
155  2020-08-31      5
156  2020-09-09      5
157  2020-01-11      6
158  2020-03-14      6
159  2020-09-02      6
160  2021-04-25      6
161  2020-01-08      7
162  2020-01-09      7
163  2020-03-19      7
164  2020-01-25      8
165  2020-02-03      8
166  2020-02-23      8
167  2020-03-21      8
168  2020-08-22      8
169  2020-08-28      8
170  2020-09-03      8
171  2020-02-15      9
172  2020-03-05      9
173 

And lastly, how many pregnant women survived/died of COVID:

In [27]:
pd.read_sql_query("""
    SELECT 
        'Deceased pregnant women' pregnant_covid, 
        COUNT(*) num_of_patients
    FROM valid_covid_19_data
    WHERE pregnant = True AND date_died IS NOT NULL
    UNION
    SELECT 
        'Alive pregnant women', 
        COUNT(*)
    FROM valid_covid_19_data
    WHERE pregnant = True and date_died IS NULL
""", engine)

pregnant_covid  num_of_patients
0     Alive pregnant women             8042
1  Deceased pregnant women               89

Thankfully, mortality among pregnant women was very low. As sad as a death is, death of an unborn child is one of the saddest things that can happen. Sadly, the data provided can't in any way answer if and how COVID affected these children, it might be interesting to see the exact results.

That's the end of my COVID-19 in Mexico case study. Hope you found it interesting! 

**Have a great day!**